contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -179,97 +222,146 @@ contract OtcOrdersFeature is
-            unwrapWeth ? WethOptions.UnwrapWeth : WethOptions.LeaveAsWeth
+            taker
         );
+
+        if (unwrapWeth) {
+            require(
+                order.makerToken == WETH,
+                "OtcOrdersFeature::fillTakerSignedOtcOrder/MAKER_TOKEN_NOT_WETH"
+            );
+            (takerTokenFilledAmount, makerTokenFilledAmount) = _settleOtcOrder(
+                order,
+                order.takerAmount,
+                taker,
+                address(this)
+            );
+            // Unwrap WETH
+            WETH.withdraw(makerTokenFilledAmount);
+            // Transfer ETH to taker
+            _transferEth(taker, makerTokenFilledAmount);
+        } else {
+            (takerTokenFilledAmount, makerTokenFilledAmount) = _settleOtcOrder(
+                order,
+                order.takerAmount,
+                taker,
+                taker
+            );
+        }
+
+        emit OtcOrderFilled(
+            orderInfo.orderHash,
+            order.maker,
+            taker,
+            address(order.makerToken),
+            address(order.takerToken),
+            takerTokenFilledAmount,
+            makerTokenFilledAmount
+        );
+    }
+
+    /// @dev Fills multiple taker-signed OTC orders.
+    /// @param orders Array of OTC orders.
+    /// @param makerSignatures Array of maker signatures for each order.
+    /// @param takerSignatures Array of taker signatures for each order.
+    /// @param unwrapWeth Array of booleans representing whether or not 
+    ///        to unwrap bought WETH into ETH for each order. Should be set 
+    ///        to false if the maker token is not WETH.
+    /// @param revertIfIncomplete If true, reverts if this function fails to
+    ///        fill any individual order.
+    /// @return takerTokenFilledAmounts Array of amounts filled, in taker token.
+    /// @return makerTokenFilledAmounts Array of amounts filled, in maker token.
+    function batchFillTakerSignedOtcOrders(
+        LibNativeOrder.OtcOrder[] memory orders,
+        LibSignature.Signature[] memory makerSignatures,
+        LibSignature.Signature[] memory takerSignatures,
+        bool[] memory unwrapWeth,
+        bool revertIfIncomplete
+    )
+        public
+        override
+        returns (
+            uint128[] memory takerTokenFilledAmounts,
+            uint128[] memory makerTokenFilledAmounts
+        )
+    {
+        require(
+            orders.length == makerSignatures.length && 
+            orders.length == takerSignatures.length &&
+            orders.length == unwrapWeth.length,
+            "OtcOrdersFeature::batchFillTakerSignedOtcOrders/MISMATCHED_ARRAY_LENGTHS"
+        );
+        takerTokenFilledAmounts = new uint128[](orders.length);
+        makerTokenFilledAmounts = new uint128[](orders.length);
+        for (uint256 i = 0; i != orders.length; i++) {
+            // Swallow reverts
+            (bool success, ) = _implementation.delegatecall(
```

Comment:
why use delegatecall instead of call here?

Reply:
@alexkroeger it needs to execute in the EP's context (which would be OK if he called `this`, but that has more overhead), and we also want to preserve `msg.sender` (taker).

Reply:
yeah just for gas savings